windows/drivers/uwb/simulator/UwbSimulatorIoEventQueue.cxx

```diff
@@ -39,83 +49,108 @@ UwbSimulatorIoEventQueue::Uninitialize()
+UwbSimulatorIoEventQueue::EnqueueRequest(WDFREQUEST request)
 {
-    NTSTATUS status = WdfWaitLockAcquire(m_wdfQueueLock, nullptr);
-    if (!NT_SUCCESS(status)) {
-        return status;
+    WDFREQUEST requestExisting = nullptr;
+    auto wdfFile = WdfRequestGetFileObject(request);
+    std::unique_lock notificationLockExclusive{ m_notificationGate };
+
+    // Check if a request is already pending.
+    NTSTATUS status = WdfIoQueueFindRequest(m_wdfQueue, nullptr, wdfFile, nullptr, &requestExisting);
+    if (requestExisting != nullptr) {
+        DbgPrint("%p an existing request already pending", m_wdfQueue);
+        WdfObjectDereference(requestExisting);
+        return STATUS_INVALID_DEVICE_STATE;
     }
 
-    if (!m_notificationQueue.empty()) {
-        auto &notificationData = m_notificationQueue.front();
-        auto converted = UwbCxDdi::From(notificationData);
-        auto outputBufferSizeRequired = converted.size();
-        if (outputBufferSize < outputBufferSizeRequired) {
-            outputBufferSize = outputBufferSizeRequired;
-            status = STATUS_BUFFER_TOO_SMALL;
-        } else {
-            notificationDataOpt = std::move(notificationData);
-            m_notificationQueue.pop();
-            status = STATUS_SUCCESS;
-        }
-    } else {
-        status = WdfRequestForwardToIoQueue(request, m_wdfQueue);
-        if (!NT_SUCCESS(status)) {
-            // TODO: log
-        }
-        status = STATUS_PENDING;
+    // Forward the request to the queue and indicate the i/o is pending.
+    DbgPrint("%p pending request %p\n", m_wdfQueue, request);
+    status = WdfRequestForwardToIoQueue(request, m_wdfQueue);
+    if (!NT_SUCCESS(status)) {
+        DbgPrint("%p failed to pend request %p with status 0x%08x\n", m_wdfQueue, status);
+        return status;
     }
 
-    WdfWaitLockRelease(m_wdfQueueLock);
+    // Notify the processing thread that a new request is pending.
+    m_notificationRequestPending.notify_one();
 
-    return status;
+    // Return pending status so dispatch code won't complete the request.
+    return STATUS_PENDING;
 }
 
-NTSTATUS
-UwbSimulatorIoEventQueue::PushNotificationData(UwbNotificationData notificationData)
+void
+UwbSimulatorIoEventQueue::PushNotification(UwbNotificationData notificationData)
 {
-    // Convert the neutral notification type to the DDI type. This needed, at
-    // minimum, to determine the required buffer size since it will eventually
-    // contain the DDI type.
-    auto notificationDataDdi = UwbCxDdi::From(notificationData);
-    auto notificationDataDdiBuffer = std::data(notificationDataDdi);
-    auto notificationDataDdiBufferSize = std::size(notificationDataDdi);
-
-    NTSTATUS status = WdfWaitLockAcquire(m_wdfQueueLock, nullptr);
-    if (!NT_SUCCESS(status)) {
-        return status;
+    // Push the notification data into the queue.
+    {
+        std::unique_lock notificationLockExclusive{ m_notificationGate };

```

Comment:
`std::lock_guard` could be used here, yes. In that case, `std::scoped_lock` is almost always the right choice instead of `std::lock_guard` since it's essentially its successor. I originally used `std::unique_lock` here because the initial implementation used a standard `std::condition_variable` which requires a `std::unique_lock` and I wanted to use consistent RAII locking objects. However, we're now using std::condition_variable_any` which takes any lockable type, not just `std::unique_lock`, so, that should take the `std::scoped_lock` as well.